### ISSUE TITLE
improper_ctypes_definitions doesn't lint on non-FFI safe fn type construction

### ISSUE THREAD

**Issue by chc4** [rust-lang/rust#88107]
Given the following code: 

```rust
#![deny(improper_ctypes_definitions)]
pub extern "C" fn third(_a: usize, b: usize, c: usize) -> usize {
    println!("arg 2 {} arg 3 {}", b, c);
    c
}


fn main() {
    let f: extern "C" fn(usize, usize, usize)->usize = third;
    let normal_third = f(1,2,3);
    assert_eq!(normal_third, 3, "normal_third");
    
    // we know that third never uses the first argument, so let's just replace it
    // it with a unit! the other two parameters should still be fine, right?
    //
    // (note that this doesn't give an improper_ctypes_definitions warning, because
    // we don't have a *definition*, only an improper extern fn type :( )
    let throw_away_f: extern "C" fn((), usize, usize)->usize =
        unsafe { std::mem::transmute(third as extern "C" fn(_,_,_)->_) };
    let throw_away_third = throw_away_f((),2,3);
    // uh-oh! the zero-sized type was actually *removed* from our systemv calling
    // convention entirely, shifting our registers over and making `c` junk!
    assert_eq!(throw_away_third, 3, "throw_away_third");
}
```
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=55030237a20e0b249ad7d977a8f6f75f)
The current output is:

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `94536864439520`,
 right: `3`: throw_away_third', src/main.rs:23:5
```

Ideally the output should look like:

```
error: `extern` fn uses type `()`, which is not FFI-safe
17 |    let throw_away_f: extern "C" fn((), usize, usize)->usize =
  |                                        ^^ not FFI-safe
```

Currently, the improper_ctypes_definitions lint will warn only for declarations of extern functions with non-FFI arguments (such as `pub extern "C" fn throw_away_third(_: (), b: usize, c: usize) -> usize { ... }`), but doesn't warn on just *using* the fn type itself, so it doesn't help in cases where you e.g. transmute a JIT function to an fn type and use it/include it in struct types itself. This kinda makes sense, in that the lint name is improper_ctypes_definitions, but my assumption when writing the code was that there was "that function FFI type safety lint" that would warn if I did something stupid without directly remembering the name.

**Comment by chc4** [comment]
(The core error in the code is the assumption that `()` would only leave the `a` value uninitialized, and not actually cause parameters to be shuffled due to the first argument size changing, which is correct behavior - I was just surprised that the lint didn't warn me about it, and it seems like an oversight.)
